### PR TITLE
Kafka insert

### PR DESCRIPTION
**Pull request by elonazoulay**
Implement insert into kafka with confluent schema registry.

### CODE REVIEW

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaMetadata.java [hunk 1]
```diff
@@ -95,6 +95,7 @@ public KafkaTableHandle getTableHandle(ConnectorSession session, SchemaTableName
                         kafkaTopicDescription.getMessage().flatMap(KafkaTopicFieldGroup::getDataSchema),
```
Comment (kokosing) [inline plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaMetadata.java:95]: PR title needs to be improved.

File: plugin/trino-kafka/pom.xml [hunk 2]
```diff
@@ -79,11 +79,21 @@
...
+        </dependency>
+
         <dependency>
             <groupId>io.confluent</groupId>
             <artifactId>kafka-schema-registry-client</artifactId>
         </dependency>
 
+        <dependency>
+            <groupId>io.confluent</groupId>
+            <artifactId>kafka-schema-serializer</artifactId>
```
Comment (kokosing) [inline plugin/trino-kafka/pom.xml:94]: What is the license of this?

File: plugin/trino-kafka/pom.xml [hunk 3]
```diff
@@ -79,11 +79,21 @@
             <artifactId>guice</artifactId>
         </dependency>
 
+        <dependency>
+            <groupId>io.confluent</groupId>
+            <artifactId>kafka-avro-serializer</artifactId>
```
Comment (kokosing) [inline plugin/trino-kafka/pom.xml:84]: What is the license of this?

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/schema/WriterSchemaResolver.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.kafka.schema;
+
+public interface WriterSchemaResolver
```
Comment (kokosing) [inline plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/schema/WriterSchemaResolver.java:16]: Why do you need such empty interface? Why not to use `ContentSchemaReader`?

**Review by kokosing — COMMENTED** [review]
We need a lot of testing:
Please consider to implement `BaseConnectorSmokeTest` where schema registry is used. So we can insert and we can select.

Try to do the same with `BaseConnectorTest`

We should remove io.trino.tests.kafka.TestKafkaAvroReadsSmokeTest.MessageSerializer and use plain INSERT instead.

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/schema/WriterSchemaResolver.java (hunk 4 above)
Comment (elonazoulay) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/schema/WriterSchemaResolver.java:16]: For the file based table description supplier a different implementation is required for the content schema reader and the file based writer schema resolver. Will look into how to unify them.

File: plugin/trino-kafka/pom.xml (hunk 2 above)
Comment (elonazoulay) [reply plugin/trino-kafka/pom.xml:94]: Apache license. Each subproject has it's own license file and all client libs use the apache license.

File: plugin/trino-kafka/pom.xml (hunk 3 above)
Comment (elonazoulay) [reply plugin/trino-kafka/pom.xml:84]: Apache license. Same applies to the kafka-schema-registry-client artifact.

**Comment by elonazoulay** [comment]
This pr uses the fix for realtime handling in the AvroColumnDecoder: #6173

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/schema/WriterSchemaResolver.java (hunk 4 above)
Comment (elonazoulay) [reply plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/schema/WriterSchemaResolver.java]: Update: I got rid of the WriterSchemaResolver, added more tests. Lmk when you have some time to take a look.

File: lib/trino-record-decoder/src/main/java/io/trino/decoder/avro/AvroColumnDecoder.java [hunk 5]
```diff
@@ -166,7 +166,7 @@ public boolean isNull()
         @Override
         public double getDouble()
         {
-            if (value instanceof Double || value instanceof Float) {
```
Comment (kokosing) [inline lib/trino-record-decoder/src/main/java/io/trino/decoder/avro/AvroColumnDecoder.java:169]: I don't think we want to remove this. This is backward incompatible change.

File: lib/trino-record-decoder/src/main/java/io/trino/decoder/avro/SingleValueRowDecoder.java [hunk 6]
```diff
@@ -23,7 +23,7 @@
 
 import static java.util.Objects.requireNonNull;
 
-class SingleValueRowDecoder
+public class SingleValueRowDecoder
```
Comment (kokosing) [inline lib/trino-record-decoder/src/main/java/io/trino/decoder/avro/SingleValueRowDecoder.java:26]: I would squash that change with the commit that is using this.

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/schema/confluent/ConfluentModule.java [hunk 7]
```diff
@@ -66,7 +68,8 @@ protected void setup(Binder binder)
         configBinder(binder).bindConfig(ConfluentSchemaRegistryConfig.class);
         install(new ConfluentDecoderModule());
         install(new ConfluentEncoderModule());
-        binder.bind(ContentSchemaReader.class).to(AvroConfluentContentSchemaReader.class).in(Scopes.SINGLETON);
+        binder.bind(AvroConfluentContentSchemaReader.class).in(SINGLETON);
+        binder.bind(ContentSchemaReader.class).annotatedWith(ForKafkaRead.class).to(AvroConfluentContentSchemaReader.class);
```
Comment (kokosing) [inline plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/schema/confluent/ConfluentModule.java:72]: Why do you bind it twice? It looks like its bounded to the same class with same dependencies twice.

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/schema/confluent/ConfluentModule.java [hunk 8]
```diff
@@ -66,7 +68,8 @@ protected void setup(Binder binder)
         configBinder(binder).bindConfig(ConfluentSchemaRegistryConfig.class);
         install(new ConfluentDecoderModule());
         install(new ConfluentEncoderModule());
-        binder.bind(ContentSchemaReader.class).to(AvroConfluentContentSchemaReader.class).in(Scopes.SINGLETON);
+        binder.bind(AvroConfluentContentSchemaReader.class).in(SINGLETON);
```
Comment (kokosing) [inline plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/schema/confluent/ConfluentModule.java:71]: Why this is needed? I don't see its usages.

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/encoder/kafka/KafkaSerializerRowEncoder.java [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+    @Override
+    protected void appendNullValue()
+    {
+        isNull = true;
+    }
+
+    @Override
+    protected void appendLong(long value)
+    {
+        data = LONG_SERIALIZER.serialize(topic, value);
```
Comment (kokosing) [inline plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/encoder/kafka/KafkaSerializerRowEncoder.java:145]: You don't append but overwrite.

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/encoder/kafka/KafkaSerializerRowEncoder.java [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+import java.util.Optional;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.trino.spi.type.TimestampType.MAX_SHORT_PRECISION;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+// Used to serialize single value keys not registered with schema registry
+public class KafkaSerializerRowEncoder
```
Comment (kokosing) [inline plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/encoder/kafka/KafkaSerializerRowEncoder.java:60]: This class should be under confluent package

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/schema/confluent/ConfluentModule.java [hunk 11]
```diff
@@ -114,6 +117,7 @@ public void configure(Binder binder)
             binder.bind(AvroDeserializer.Factory.class).to(AvroBytesDeserializer.Factory.class).in(Scopes.SINGLETON);
             newMapBinder(binder, String.class, RowDecoderFactory.class).addBinding(AvroRowDecoderFactory.NAME).to(AvroRowDecoderFactory.class).in(Scopes.SINGLETON);
             newMapBinder(binder, String.class, RowDecoderFactory.class).addBinding(DummyRowDecoder.NAME).to(DummyRowDecoderFactory.class).in(SINGLETON);
+            newMapBinder(binder, String.class, RowDecoderFactory.class).addBinding(KafkaRowDecoderFactory.NAME).to(KafkaRowDecoderFactory.class).in(SINGLETON);
```
Comment (kokosing) [inline plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/schema/confluent/ConfluentModule.java:120]: I was expecting to use `AvroRowDecoderFactory`. Why do you have some custom decoder here? I was expecting that insert will write avro data.

File: plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaTableHandle.java [hunk 12]
```diff
@@ -54,6 +54,7 @@
     private final Optional<String> messageDataSchemaLocation;
     private final Optional<String> keySubject;
     private final Optional<String> messageSubject;
+    private final Optional<List<String>> keyColumns;
```
Comment (kokosing) [inline plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaTableHandle.java:57]: I don't see where this field is used. Can you please help me?

**Review by kokosing — COMMENTED** [review]
Started reading this. I don't understand why you need custom serialization and deserialization.

File: lib/trino-record-decoder/src/main/java/io/trino/decoder/avro/AvroColumnDecoder.java (hunk 5 above)
Comment (kokosing) [reply lib/trino-record-decoder/src/main/java/io/trino/decoder/avro/AvroColumnDecoder.java:169]: Oh.... I see it comes from https://github.com/trinodb/trino/pull/6173

**Comment by bitsondatadev** [comment]
:wave: @elonazoulay - this PR is inactive and doesn't seem to be under development. If you'd like to continue work on this at any point in the future, feel free to re-open.

**Pull request closed without merging**